### PR TITLE
[CR] Randomize test order: catch interaction, etc bugs

### PR DESCRIPTION
**Pull request by actual-nh**
#### Summary

SUMMARY: Infrastructure "Randomize test order to catch interactions, other bugs"

#### Purpose of change

As seen in #46439 and #46404, some test faults are triggered by differing test orders. Once this and other existing faults are fixed using data from testing this PR, it can be merged to help with future testing.

#### Describe the solution

Add `--order rand` when invoking cata_test.

#### Describe alternatives you've considered

I can't see any practical ones.

#### Testing

See #46439.

#### Additional context

I plan on using the results from this PR to fix the tests, test them on this PR, then put in other PRs for the individual fixes. (Unless someone has a better idea?)

### CODE REVIEW

**Comment by actual-nh** [comment]
I may do a PR to note in TESTING.md exactly where one needs to do to globally modify what cata_test (or various alternate names) does...

**Comment by actual-nh** [comment]
I have asked in Discussions (#46476) if someone with a Linux box can analyze the Github "artifacts".

**Comment by actual-nh** [comment]
Current status:

- Patch to add caloric replenishment to clear_avatar() has succeeded in eliminating starvation of the test subject. However, it may be altering weary test results a bit - will take a further look at thresholds, etc to see what's happening and what to do.
- The manhack release test is failing because there isn't any room to do so. I will check if this is lacking a clear_map() before it.

**Comment by actual-nh** [comment]
Currently:
- Several involving time (and light levels). Possibly something is leaving a modified value for season length?
- The biometrics test may be making assumptions as to how much activity the avatar has done earlier; not sure, though.
- Vehicle level test keeps erroring (no surprise; #46441).
- Weary checks are still erroring (original may have had the level going up and down way too much, however), with the weary threshold still rather unstable (see #46384 for a bit more on this - may add more output to debug_weary_info to start investigating this more).

**Comment by actual-nh** [comment]
May want to alter the various cata_test calls to have `--durations yes` while doing `--order rand`, to make sure know what's before and after what.

**Comment by actual-nh** [comment]
As it turns out, monster_test.cpp was altering `trigdist` and not putting it back to the same value. (A REQUIRE check in `vision_single_tile_skylight` for `trigdist == true` may be indicated also.)

File: tests/monster_test.cpp [hunk 1]
```diff
@@ -300,37 +300,57 @@ static void monster_check()
 TEST_CASE( "write_slope_to_speed_map_trig", "[.]" )
 {
     clear_map_and_put_player_underground();
+    bool old_trigdist = trigdist;
     override_option opt( "CIRCLEDIST", "true" );
     trigdist = true;
     test_moves_to_squares( "mon_zombie_dog", true );
     test_moves_to_squares( "mon_pig", true );
+    if( ! old_trigdist ) {
+        override_option opt( "CIRCLEDIST", "false" );
```
Comment (anothersimulacrum) [inline tests/monster_test.cpp:309]: I'm pretty sure this won't work, it'll immediately go out of scope and have no effect.

Comment (actual-nh) [reply tests/monster_test.cpp:309]: So just set trigdist back to old_trigdist?

Comment (anothersimulacrum) [reply tests/monster_test.cpp]: > So once that variable goes out of scope, the destructor puts that option (as opposed to trigdist!) back to the old version?

Correct.

**Comment by actual-nh** [comment]
So, errors still seen:

- Weary_recovery and/or weary_24h_tasks - virtually always at least one; usually both; at least 50% have weird fluctuations in weariness levels and/or weariness threshold (#46384)
- Vehicle_level_test (#46441) - usually seen
- All_nutrition_starve_test (for vitamin C) - sometimes seen
- Intermittently a few others like item name tests and a field test (#46256 shows other examples)

**Comment by actual-nh** [comment]
The expanded debug_weary_info has revealed that the `weary_recovery` vehicle work results in a fatigue of -1000 (any lower gets automatically reset to this), thus explaining the very high threshold (high fatigue causes a lower threshold, and the reverse is likewise true). I've tried resetting -20 or below (the point at which the character would otherwise wake up) fatigue to 0 in `activity_scheduling_helper.cpp`, but without success. I now think the solution is to place a minimum value of -20 (fatigue) on fatigue's effects on weary_threshold.

File: src/character.cpp [hunk 2]
```diff
@@ -4972,9 +4972,13 @@ std::string Character::debug_weary_info() const
...
     int current = weariness_level();
+    int morale = get_morale_level();
+    int weight = units::to_gram<int>( bodyweight() );
+    float bmi = get_bmi();
 
-    return string_format( "Weariness: %s Max Exertion: %s Mult: %g\nBMR: %d Intake: %d Tracker: %d Thresh: %d At: %d\nCalories: %d",
-                          amt, max_act, move_mult, bmr, intake, input, thresh, current, stored_calories );
+    return string_format( "Weariness: %s Max Full Exert: %s Mult: %g\nBMR: %d Intake: %d Tracker: %d Thresh: %d At: %d\nCal: %d/%d Fatigue: %d Morale: %d Wgt: %d (BMI %.1f)",
+                          amt, max_act, move_mult, bmr, intake, input, thresh, current, stored_calories,
+			              healthy_calories, fatigue, morale, weight, bmi );
```
Comment (actual-nh) [inline src/character.cpp:4981]:
```suggestion
                          amt, max_act, move_mult, bmr, intake, input, thresh, current, stored_calories,
                          healthy_calories, fatigue, morale, weight, bmi );
```
To me, these were already at the same level. Tabs?

File: tests/activity_scheduling_helper.cpp [hunk 3]
```diff
@@ -196,8 +202,9 @@ std::string weariness_events::summarize() const
 {
     std::string buffer;
     for( const weary_transition &change : transitions ) {
-        buffer += string_format( "Transition: Weariness from %d to %d at %d minutes\n", change.from,
-                                 change.to, change.minutes );
+        buffer += string_format( "Transition: Weariness lvl from %d to %d at %d min (W %d Th %d)\n",
+				                 change.from, change.to, change.minutes,
+				                 change.new_weariness, change.new_threshold );
```
Comment (actual-nh) [inline tests/activity_scheduling_helper.cpp:207]:
```suggestion
        buffer += string_format( "Transition: Weariness lvl from %d to %d at %d min (W %d Th %d)\n",
				                                  change.from, change.to, change.minutes,
				                                  change.new_weariness, change.new_threshold );
```

File: tests/activity_scheduling_helper.h [hunk 4]
```diff
@@ -93,12 +93,15 @@ struct weariness_events {
...
+	        int new_weariness = 0;
+	        int new_threshold = 0;
         };
 
         std::vector<weary_transition> transitions;
 
     public:
-        void log( int old_level, int new_level, const time_duration &when );
+        void log( int old_level, int new_level, const time_duration &when, int new_weariness,
+		  int new_threshold );
```
Comment (actual-nh) [inline tests/activity_scheduling_helper.h:104]:
```suggestion
        void log( int old_level, int new_level, const time_duration &when, int new_weariness,
		                int new_threshold );
```

**Review by actual-nh — COMMENTED** [review]
Half the time, I don't see any differences.

**Comment by actual-nh** [comment]
I have no idea what's up with astyle, or Github, in terms of spacing. On my local machine, it looks fine; when it looks wrong on Github, and I do a comment that should fix it, half the time it comes out with something else that's wrong - but looked fine in the suggested change in the comment.

**Comment by actual-nh** [comment]
Note that the g++-9 did actually get some tests in before cancellation. (Regarding cancellation - I have to wonder if there's a way to still cache what useful object files were produced prior to the cancellation.) @kevingranade: Both it and the g++-7 had failures for the dodge testing (and I made sure to pull in CleverRaven/CDDA:master after your changes in that regard.

**Comment by actual-nh** [comment]
BTW, I'm seeing if a local astyle + making sure git won't allow through initial tabs will help.

**Comment by actual-nh** [comment]
I think I'm going to need to do a PR to put the extra `weary` debugging output into the master branch. That would be a more efficient/fair way to get more test data than, say, changing the setup for this branch so it didn't cancel any Github test builds.

File: tests/player_helpers.cpp [hunk 5]
```diff
@@ -72,7 +72,9 @@ void clear_character( player &dummy )
     // This sets HP to max, clears addictions and morale,
     // and sets hunger, thirst, fatigue and such to zero
     dummy.environmental_revert_effect();
-    dummy.set_stored_kcal( dummy.get_healthy_kcal() ); // But not stored kcal
+    // However, the above does not set stored kcal;
+    // 2170 is calories of debug_nutrition
```
Comment (anothersimulacrum) [inline tests/player_helpers.cpp:76]: Stick a `REQUIRE` or something here guaranteeing that?

Comment (actual-nh) [reply tests/player_helpers.cpp:76]: Point - but it's erroring in stomach contents bmr requirements, so I may need to revert that anyway.

**Comment by actual-nh** [comment]
The Appveyor results from the above are the most informative regarding weary; Travis also has one of immediate interest, although approximately duplicated in Appveyor; general is from the mixed-work/food (and other non-weary), which is harder to interpret. I'm going to try removing the -2170 (and also do a slight bit of reformatting so messages don't wrap) next, somewhat to confirm these results (from which it appears the `intake` is what's fluctuating weariness up and down...).

**Comment by actual-nh** [comment]
The general, travis, and appveyor test failures are indicating that the 24-hour-dig task's fluctuating from 1 to 0 to 1 in weariness level is due to intake going up (as calories are absorbed from the guts). They unfortunately do not give information for the 8-hours-dig, 8-hours-wait A local test using `--success` to get full test output is yielding about the same for the 24-hour-dig task, but the 8-hours/8-hours task seems to be fluctuating due to more interactions between tracker and intake:
```
Digging Pits 8 hours, then waiting 8:
  Weariness: 0 Max Full Exert: EXTRA_EXERCISE Mult: 1
  BMR: 1738 Intake: 0 Tracker: 0 Thresh: 1038 At: 0
  Cal: 55000/55000 Fatigue: 0 Morale: 0 Wgt: 76562 (BMI 25.0)
  Transition: Weary lvl 0 to 1 at 120 min (W 1028 Th 1020 Tr 1450 In 843)
  Transition: Weary lvl 1 to 2 at 250 min (W 2025 Th 1000 Tr 2702 In 1353)
  Transition: Weary lvl 2 to 3 at 360 min (W 2740 Th 983 Tr 3494 In 1507)
  Transition: Weary lvl 3 to 4 at 465 min (W 3222 Th 968 Tr 3997 In 1549)
  Transition: Weary lvl 4 to 3 at 520 min (W 3128 Th 960 Tr 3889 In 1521)
  Transition: Weary lvl 3 to 4 at 540 min (W 3181 Th 957 Tr 3913 In 1463)
  Transition: Weary lvl 4 to 3 at 550 min (W 2997 Th 956 Tr 3729 In 1463)
  Transition: Weary lvl 3 to 2 at 640 min (W 2581 Th 944 Tr 3296 In 1429)
  Transition: Weary lvl 2 to 3 at 650 min (W 2593 Th 943 Tr 3308 In 1429)
  Transition: Weary lvl 3 to 2 at 670 min (W 2483 Th 940 Tr 3166 In 1366)
  Transition: Weary lvl 2 to 1 at 875 min (W 1815 Th 911 Tr 2418 In 1205)
  Transition: Weary lvl 1 to 2 at 885 min (W 1827 Th 910 Tr 2430 In 1205)
  Transition: Weary lvl 2 to 1 at 905 min (W 1757 Th 907 Tr 2332 In 1149)
  Weariness: 1733 Max Full Exert: ACTIVE_EXERCISE Mult: 0.8
  BMR: 1722 Intake: 1093 Tracker: 2280 Thresh: 900 At: 1
  Cal: 52525/55000 Fatigue: 193 Morale: 0 Wgt: 74908 (BMI 24.5)
```
I suspect the above is due to `intake` being subtracted every 12 times `try_reduce_weariness` is called, while `tracker` can be subtracted up to 4 times that frequently (if sleeping; up to 2 times that frequently if NO_EXERCISE, which should be the case after 480 minutes here). Some of this may be relieved by #45316, but not the 24-hour-dig test problem. This may also be helped by tracking caloric intake over a longer time-frame (say, 24 hours; note #36976).

However... @I-am-Erk, @anothersimulacrum: I am unclear on the exact purpose of factoring caloric intake into weariness. (Currently, one would almost say the character has hypoglycemia...)

**Comment by jbytheway** [comment]
Really glad to see someone working on getting randomly ordered tests working.  It's something that was on my todo list but I probably never would have got around to.

I just want to make sure you're aware of `tools/reduce_tests.sh`.  It's really helpful for figuring out the source of inter-test dependencies (though, if you're on Windows I'm not sure how much help it would be.  You might still be able to use it via one of the various Windows shells).

If there are specific things you'd like to see tested on Linux, let me know, and I'll try to get to them.

**Comment by actual-nh** [comment]
> Really glad to see someone working on getting randomly ordered tests working. It's something that was on my todo list but I probably never would have got around to.

Understand! Things are about to get busier for me at work (teaching Anatomy & Physiology), but I'll do as much as I can. (Once the parallel tests stabilize a bit, I will see about adapting it for randomized test orders - should not be difficult; sticking `--order rand` into EXTRA_TEST_OPTS seems the best to adapt to this and any other changes.)

> I just want to make sure you're aware of `tools/reduce_tests.sh`. It's really helpful for figuring out the source of inter-test dependencies (though, if you're on Windows I'm not sure how much help it would be. You might still be able to use it via one of the various Windows shells).

Thanks! I am on a Mac, so that works; will take a look at now.

> If there are specific things you'd like to see tested on Linux, let me know, and I'll try to get to them.

Thanks!

**Comment by actual-nh** [comment]
@anothersimulacrum: Feel free to put a `Feature Freeze` on this one.

**Comment by Night-Pryanik** [comment]
I suppose this PR is abandoned, since last commit was a year and a half ago? I'm not sure if it should stay in PR tracker, especially with this many unresolved conflicts.

**Comment by Night-Pryanik** [comment]
Closing as abandoned.

**Pull request closed without merging**